File: itg/code/itg.c
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>

#define analysis_main main

/******************************************************************************
 * Common helpers
 ******************************************************************************/

/*
   1  2  3
      U
   4 L R 5
      D
   6  7  8

 Figure 1. The Roxor, Inc. In The Groove Dance Game (henceforth "ITG") Dance
 Step Direction Indicators (henceforth "Arrows"), UDLR, and the 8 possible
 directions of facing.
 */

enum facing {
	F_UL, F_U, F_UR,
	F_L,  F_N, F_R,
	F_DL, F_D, F_DR,
	NUM_FACINGS,
};

enum arrow { U, D, L, R, NUM_ARROWS };

enum facing whichway(enum arrow left_foot, enum arrow right_foot) {
	// Foot switches: future work!
	assert(left_foot != right_foot);
	static enum facing way_table[NUM_ARROWS][NUM_ARROWS] = {
		[U][D] = F_R,
		[U][L] = F_DR, // xover
		[U][R] = F_UR,
		//[U][U] = F_U, // footswitch
		[L][U] = F_UL,
		[L][D] = F_UR,
		[L][R] = F_U,
		//[L][L] = F_L, // xover footswitch
		[D][U] = F_L,
		[D][L] = F_DL, // xover
		[D][R] = F_UL,
		//[D][D] = F_U, // footswitch (to be theoretically beautiful, this would be F_D, a backwards switch, but who the hell does that?)
		[R][L] = F_D, // lateral
		[R][U] = F_DL, // xover
		[R][D] = F_DR, // xover
		//[R][R] = F_R, // xover footswitch
	};
	return way_table[left_foot][right_foot];
}

// For any two facings possible with consecutive itg steps (ie, only 1 foot
// moves), shouldn't return more than 2.
// Higher turniness is possible in future work using jumps!
unsigned int turn_distance(enum facing before, enum facing after)
{
	assert(before < NUM_FACINGS && after < NUM_FACINGS);
	static int turn_table[NUM_FACINGS][NUM_FACINGS] = {
		[F_UL] = { 0, 1, 2,
			   1, 0, 3,
			   2, 3, 4, },
		[F_U]  = { 1, 0, 1,
			   2, 0, 2,
			   3, 4, 3, },
		[F_UR] = { 2, 1, 0,
			   3, 0, 1,
			   4, 3, 2, },
		[F_L]  = { 1, 2, 3,
			   0, 0, 4,
			   1, 2, 3, },
		[F_N]  = { 0, 0, 0,
			   0, 0, 0,
			   0, 0, 0, }, // special case facing used when processing jumps - handling jump facing is future work
		[F_R]  = { 3, 2, 1,
			   4, 0, 0,
			   3, 2, 1, },
		[F_DL] = { 2, 3, 4,
			   1, 0, 3,
			   0, 1, 2, },
		[F_D]  = { 3, 4, 3,
			   2, 0, 2,
			   1, 0, 1, },
		[F_DR] = { 4, 3, 2,
			   3, 0, 1,
			   2, 1, 0, },
	};
	return turn_table[before][after];
}

/******************************************************************************
 * Analysis
 ******************************************************************************/

// 1 = tap; 2 = hold; 3 = hold release; 4 = roll; M = mine
// IDK what stepmania does for lifts/fakes; I don't know of any chart that uses them anyway.
#define COULD_BE_NOTE(c) (((c) >= '0' && (c) <= '9') || (c) == 'M' || (c) == 'm')
#define IS_NOTE(c) ((c) == '1' || (c) == '2' || (c) == '4')

bool get_note(char *buf, enum arrow *note)
{
	char l, d, u, r;
	if (sscanf(buf, "%c%c%c%c\n", &l, &d, &u, &r) == 4) {
		if (COULD_BE_NOTE(l) && COULD_BE_NOTE(d) && COULD_BE_NOTE(u) && COULD_BE_NOTE(r)) {
			if (IS_NOTE(l) && !IS_NOTE(d) && !IS_NOTE(u) && !IS_NOTE(r)) {
				*note = L;
				return true;
			} else if (!IS_NOTE(l) && IS_NOTE(d) && !IS_NOTE(u) && !IS_NOTE(r)) {
				*note = D;
				return true;
			} else if (!IS_NOTE(l) && !IS_NOTE(d) && IS_NOTE(u) && !IS_NOTE(r)) {
				*note = U;
				return true;
			} else if (!IS_NOTE(l) && !IS_NOTE(d) && !IS_NOTE(u) && IS_NOTE(r)) {
				*note = R;
				return true;
			}
		}
	}
	return false;
}

void analysis_main()
{
	char buf[256];
	enum facing last_facing;
	enum arrow last_note;
	unsigned int notes = 0;
	unsigned int turniness = 0;
	while (fgets(buf, 256, stdin)) {
		buf[255] = 0;
		enum arrow note;
		// XXX: Ignores jumps, hands, &c.
		if (get_note(buf, &note)) {
			if (notes != 0) {
				// Check for double-step. Foot switches are future work!
				if (note == last_note) {
					continue;
				}
				// XXX: Assumes no doublesteps.
				// NB. Assumes song always starts on left foot.
				// But tbh, it doesn't affect the turniness calculation
				// if we think you're backwards the entire song!
				// XXX: Fails to parse the Boten Anna roll-stream pattern.
				enum facing this_facing = notes % 2 == 0 ?
					whichway(note, last_note) :
					whichway(last_note, note);
				if (notes > 1) {
					// subsequent notes not first or second ones
					unsigned int distance = turn_distance(last_facing, this_facing);
					assert(distance <= 2);
					turniness += distance;
				}
				last_facing = this_facing;
			}
			last_note = note;
			notes++;
		}
	}
	printf("%u turniness (x10000) (%u turns / %u notes)\n", (turniness * 10000 / notes), turniness, notes);
}


/******************************************************************************
 * Synthesis
 ******************************************************************************/

bool is_270(enum facing new_facing, enum facing last_facing, enum facing two_facings_ago) {
	// Facing F_D is allowed, but twisting past that to face the opposite direction you came
	// from isn't.
	if (new_facing == F_DR) {
		if (last_facing == F_DL) {
			return true; // LURD
		} else if (last_facing == F_D && two_facings_ago == F_DL) {
			return true; // LRDLRD
		}
	// symmetric case
	} else if (new_facing == F_DL) {
		if (last_facing == F_DR) {
			return true; // RULD
		} else if (last_facing == F_D && two_facings_ago == F_DR) {
			return true; // RLDRLD
		}
	}
	return false;
}
bool is_spin(enum facing new_facing, enum facing last_facing, enum facing two_facings_ago, enum facing three_facings_ago) {
	// Patterns such as LURD are allowed but you have to twist back around the way you came
	// from. So check at facings L, R, UR, and UL.
	return (new_facing == F_R || new_facing == F_UR ||
		new_facing == F_L || new_facing == F_UL) &&
		is_270(last_facing, two_facings_ago, three_facings_ago);
}

#define SEARCHDEPTH 18
#define MAX(x, y) ({ typeof(x) __x = (x); typeof(y) __y = (y); __x < __y ? __y : __x; })

enum search_rule { NONE = 0, NO_SPINS, NO_270S, NO_LATERALS, NO_XOVERS, NUM_RULES };

void print(enum arrow *chart, unsigned int turniness)
{
	printf("turniness %u: ", turniness);
	for (unsigned int i = 0; i < SEARCHDEPTH; i++) {
		printf("%c", chart[i] == L ? 'L' : chart[i] == D ? 'D' : chart[i] == U ? 'U' : chart[i] == R ? 'R' : '?');
	}
	printf("\n");
}

// returns max turniness reachable
unsigned int search(enum arrow *chart, unsigned int index, unsigned int turniness_so_far, enum search_rule rule)
{
	unsigned int max_turniness = 0;
	if (index == 0 || index == 1) {
		// no previous facing; just do all possible 2-step starts
		for (enum arrow l = 0; l < NUM_ARROWS; l++) {
			for (enum arrow r = 0; r < NUM_ARROWS; r++) {
				if (r == l) continue;
				// avoid illegal starting positions
				if (r == L || l == R) continue;
				// other possible approach
				// if (rule >= NO_LATERALS && whichway(l,r) == F_D) continue;
				// if (rule >= NO_XOVERS && (whichway(l,r) == F_DL || whichway(l,r) == F_DR)) continue;

				// WLOG we say start on left foot
				chart[0] = l;
				chart[1] = r;
				max_turniness = MAX(max_turniness, search(chart, 2, 0, rule));
			}
		}
	} else if (index == SEARCHDEPTH) {
		print(chart, turniness_so_far);
		return turniness_so_far;
	} else {
		// general case
		for (enum arrow next = 0; next < NUM_ARROWS; next++) {
			if (next == chart[index-1]) continue; // footswitches are future work
			// which foot?
			enum facing new_facing;
			enum facing last_facing;
			enum facing two_facings_ago;
			enum facing three_facings_ago;
			if (index % 2 == 0) {
				// About to place a left foot step at chart[index].
				new_facing        =                    whichway(next,           chart[index-1]);
				last_facing       =                    whichway(chart[index-2], chart[index-1]);
				two_facings_ago   = index <= 2 ? F_N : whichway(chart[index-2], chart[index-3]);
				three_facings_ago = index <= 3 ? F_N : whichway(chart[index-4], chart[index-3]);
			} else {
				// Right foot step at chart[index].
				new_facing        =                    whichway(chart[index-1], next);
				last_facing       =                    whichway(chart[index-1], chart[index-2]);
				two_facings_ago   = index <= 2 ? F_N : whichway(chart[index-3], chart[index-2]);
				three_facings_ago = index <= 3 ? F_N : whichway(chart[index-3], chart[index-4]);
			}
			// Consider the sequence of facings. Does the search rule allow it?
			if (rule >= NO_XOVERS) {
				if (new_facing == F_DL || new_facing == F_DR) continue;
			}
			if (rule >= NO_LATERALS) {
				if (new_facing == F_D) continue;
			}
			if (rule >= NO_270S) {
				if (is_270(new_facing, last_facing, two_facings_ago)) continue;
			}
			if (rule >= NO_SPINS) {
				if (is_spin(new_facing, last_facing, two_facings_ago, three_facings_ago)) continue;
			}
			// Got here.
			chart[index] = next;
			unsigned int new_turniness = turniness_so_far + turn_distance(last_facing, new_facing);
			max_turniness = MAX(max_turniness, search(chart, index+1, new_turniness, rule));
		}
	}
	return max_turniness;
}

void synthesis_main()
{
	enum arrow chart[SEARCHDEPTH];
	unsigned int result = search(chart, 0, 0, NONE); // can customize this
	printf("Best turniness %u\n", result);
}

File: itg/code/ITG.hs
import Data.List
import Data.Char
import Data.Maybe
import Data.Tuple
import Debug.Trace

data Arrow = L| D | U | R deriving (Ord, Eq, Show)

data Jump = LD | LU | DR | UR | LR | DU | Other deriving (Ord, Eq, Show)

data Step = A Arrow | J Jump deriving (Ord, Eq, Show)

data Foot = LeftFoot | RightFoot deriving (Ord, Eq, Show)

alternate LeftFoot = RightFoot
alternate RightFoot = LeftFoot

-- TODO add a field to track what the last arrow the left foot and right foot were on
-- if you commitstream flipped, you need to flip those around too
data AnalysisState = S { steps :: Int, xovers :: Int, switches :: Int,
                         jacks :: Int, doubles :: Int, xoverfs :: Int,
                         brackets :: Int, stepsAndJumps :: Int,
                         lastStep :: Maybe Arrow, lastJack :: Maybe Arrow,
                         lastFlip :: Bool, lastFoot :: Maybe Foot,
                         justBracketed :: Bool, stepsLR :: [Bool],
                         -- used for bracket jumps e.g. if left foot's last
                         -- step was on U, right foot cannot bracket UR
                         lastArrowLR :: ([Arrow], [Arrow]), -- (left, right)
                         lastTrueLastArrowLR :: ([Arrow], [Arrow]) -- ugh...
                       } deriving Show

commitStream tiebreakFoot s = s' -- traceShow ("commit", s, "=>", s') s'
    where s' = commitStream' tiebreakFoot s

commitStream' :: Maybe Foot -> AnalysisState -> AnalysisState
commitStream' tiebreakFoot s = maybe s0 splitStream splitIndex
    where -- number of steps in the chunk of stream
          ns = length $ stepsLR s
          -- number of (seemingly) crossed-over steps
          nx = length $ filter not $ stepsLR s
          -- if more than half the L/R steps in this stream were crossed over,
          -- then we got the footing backwards and need to flip the stream.
          -- as tiebreaker #1, don't flip if immediately preceded by a bracket
          flipTieBreak _ | justBracketed s = False
          -- as tiebreaker #2, if the stream is followed by a bracketable jump,
          -- choose whichever flip-or-no-flip lets us bracket that jump.
          flipTieBreak (Just bracketFoot) = lastFoot s == Just bracketFoot
          -- as tiebreaker #3, flip if the chart is already more jacky than
          -- footswitchy, i.e., if past streams flipped more often than not.
          flipTieBreak Nothing = (switches s > jacks s) == lastFlip s
          f = nx * 2 > ns || nx * 2 == ns && flipTieBreak tiebreakFoot
          -- if "too much" of the stream is *completely* crossed-over, force
          -- a double-step there by splitting the stream to stay facing forward.
          -- heuristic value was chosen by inspection on Subluminal - After Hours.
          splitIndex = findIndex (isPrefixOf $ replicate 9 f) $ tails $ stepsLR s
          -- prevent infinite splittage if the fux section starts immediately;
          -- in that case split instead at fux's end (i.e., first non-crossed step).
          splitStream 0 = splitStream $ fromJust $ findIndex (/= f) $ stepsLR s
          splitStream i = s2 { doubles = doubles s2 + 1 }
              where (steps1, steps2) = splitAt i $ stepsLR s
                    s1 = commitStream Nothing      s  { stepsLR = steps1 }
                    s2 = commitStream tiebreakFoot s1 { stepsLR = steps2, lastJack = Nothing }
          -- normal case; consider the whole stream at once
          jack = lastJack s /= Nothing
          jackLR = lastJack s == Just L || lastJack s == Just R
          -- this nonsense merges the lastArrow lists with a known good state
          -- snapshotted from the previous stream commit, because a 1-step long
          -- stream e.g. may end up with an empty list on the other foot, and
          -- you want neither to clobber the old value with [], or clobber it by
          -- writing willy nilly while you're still not sure if you'll flip.
          -- analyze doesn't need to worry about this; the correct thing will
          -- show up in lastArrowLR each time regardless.
          unifySideArrows []     old = old
          unifySideArrows recent old = recent
          unifyArrows (ls,rs) = (unifySideArrows ls $ fst $ lastTrueLastArrowLR s,
                                 unifySideArrows rs $ snd $ lastTrueLastArrowLR s)
          -- update the true lastFoot and lastArrows (if nonzero stream)
          -- this way the value always shows our true position on the pad after
          -- the stream ended, to see if bracket jumps are possible
          (trueLastFoot, trueLastArrowLR) =
              -- was flipped; flip these (even if ns == 0, b/c that excludes U/D)
              if f then (alternate <$> lastFoot s, unifyArrows $ swap $ lastArrowLR s)
              -- not flipped: keep as is
              else (lastFoot s, unifyArrows $ lastArrowLR s)
          s0 = s { xovers   = xovers   s + if f then ns - nx else nx,
                   switches = switches s + fromEnum (f == lastFlip s && jack),
                   jacks    = jacks    s + fromEnum (f /= lastFlip s && jack),
                   xoverfs  = xoverfs  s + fromEnum (f == lastFlip s && jackLR),
                   lastFoot = trueLastFoot, lastArrowLR = trueLastArrowLR,
                   lastTrueLastArrowLR = trueLastArrowLR, -- update snapshot
                   justBracketed = False,
                   -- if we had to flip a stream right after a bracket jump,
                   -- that'd make it retroactively unbracketable; if so cancel it
                   brackets = brackets s - (if f && justBracketed s && ns > 0 then 1 else 0),
                   lastFlip = f, stepsLR = [] }

-- a jump resets the footing, so the next step can be stepped with either
-- foot. commit the stream so far to treat it separately from what follows.
-- this is the old version; still holds for LR and DU jumps, as well as 3+s.
-- we clear lastArrowLR here to represent resetting footing, and also clear the
-- lastFoot; if it's bracketable, that will be re-set afterwards down below.
commitJump s = s { lastStep = Nothing, lastJack = Nothing,
                   lastFoot = Nothing, justBracketed = False,
                   stepsAndJumps = stepsAndJumps s + 1,
                   lastArrowLR = ([],[]), lastTrueLastArrowLR = ([],[]) }

analyzeStep :: AnalysisState -> Step -> AnalysisState
analyzeStep s (J DU) =
    -- for DU jumps, try to mimimic the facing the player currently has
    -- e.g. if L D DU LU, the DU faces right (left on U) so the LU can bracket
    (commitJump s') { lastArrowLR = newLastArrows, lastTrueLastArrowLR = newLastArrows }
    where s' = commitStream Nothing s
          -- there's gotta be a cleaner way to do this?
          decideDUFacing' _ False False True = ([D],[U]) -- easy case
          decideDUFacing' True False False _ = ([D],[U]) -- easy case
          decideDUFacing' False _ True False = ([U],[D]) -- easy case
          decideDUFacing' False True _ False = ([U],[D]) -- easy case
          decideDUFacing' True False True False = ([],[]) -- D footswitch; allow anyth
          decideDUFacing' False True False True = ([],[]) -- U footswitch; allow anyth
          decideDUFacing' False False False False = ([],[]) -- no footing known
          decideDUFacing' _ _ _ _ = error "same foot on both D and U at once??"
          decideDUFacing a b c d = x -- traceShow ("deciding ud facing",a,b,c,d,x) x
              where x = decideDUFacing' a b c d
          newLastArrows = decideDUFacing (elem D $ fst $ lastArrowLR s') -- left on D?
                                         (elem U $ fst $ lastArrowLR s') -- or on U?
                                         (elem D $ snd $ lastArrowLR s') -- right on D?
                                         (elem U $ snd $ lastArrowLR s') -- or on U?
analyzeStep s (J jump) | elem jump [LR,Other] =
    let s'' = commitJump $ commitStream Nothing s in s'' -- traceShow ("other jump", s, s'') s''
analyzeStep s (J jump) =
    if can_bracket jump newLastFoot lastArrows then
        (commitJump s') { brackets = brackets s' + 1,
                          lastFoot = Just newLastFoot,
                          justBracketed = True, -- contingent on not next stream flip
                          lastArrowLR = ([],[]), -- newLastArrows,
                          lastTrueLastArrowLR = newLastArrows -- XXX: is this correct? seems to work without
                        }
    else
        let s'' = commitJump s' in s'' -- traceShow ("unbrackerable jump", s, s'') s'' -- jump it normally
    -- TODO: instruct commitstream to breka ties to bias towards bracketing this
    -- TODO-2: do a "check" for bracket on subsequent stream where you fix the
    -- foot and test if it flips, and if it needs to flip, you cancel the bracket
    -- ...either that, or you actually FORCE it never to flip
    -- idk which would be more natural!
    where s' = commitStream (Just $ defaultJumpFoot jump) s -- tell it how to break tie
          defaultJumpFoot LD = LeftFoot
          defaultJumpFoot LU = LeftFoot
          defaultJumpFoot DR = RightFoot
          defaultJumpFoot UR = RightFoot
          -- if foot-alternation is not established, e.g. we just did a LR jump,
          -- or it's the very start of the chart, err towards bracketing
          -- XXX: what is this doing here?
          -- XXX: the foot should either be the bracket foot, or Nothing at all
          newLastFoot = maybe (defaultJumpFoot jump) alternate $ lastFoot s'
          lastArrows = lastArrowLR s'
          bracketArrows LD = [L,D]
          bracketArrows LU = [L,U]
          bracketArrows DR = [D,R]
          bracketArrows UR = [U,R]
          -- test if e.g. right foot is on U, then left foot cannot bracket LU
          interferes footArrows j = any (\a -> elem a footArrows) $ bracketArrows j
          -- note 2nd argument is the foot to attempt to bracket with
          can_bracket' LD LeftFoot (_, lastRight) = not $ interferes lastRight LD
          can_bracket' LU LeftFoot (_, lastRight) = not $ interferes lastRight LU --(elem L lastRight) && not (elem U lastRight)
          can_bracket' DR RightFoot (lastLeft, _) = not $ interferes lastLeft DR
          can_bracket' UR RightFoot (lastLeft, _) = not $ interferes lastLeft UR
          can_bracket' _ _ _ = False -- don't allow jack-brackets (future work?)
          can_bracket j foot (ll,lr) =
              -- traceShow ("can bracker?",j,foot,ll,lr) $ can_bracket' j foot (ll,lr)
              can_bracket' j foot (ll,lr)
          -- again the foot is the foot we bracket with
          updateLastArrow LeftFoot  (lastLeft,lastRight) arrows = (arrows,lastRight)
          updateLastArrow RightFoot (lastLeft,lastRight) arrows = (lastLeft,arrows)
          -- compute new foot positions, persisting thru the commit-jump
          newLastArrows = updateLastArrow newLastFoot lastArrows $ bracketArrows jump
          
analyzeStep s (A arrow)
    -- two steps on the same arrow might be a jack, or might be a footswitch.
    -- to figure out which, commit the stream so far, and begin a new stream
    -- whose footing will retroactively determine how to foot this step.
    -- also, unlike jumps, this step gets counted as part of the next stream.
    | lastStep s == Just arrow = stream (commitStream Nothing s) { lastJack = Just arrow }
    -- a normal streamy step.
    | otherwise = stream s
    where foot = maybe LeftFoot alternate $ lastFoot s
          -- record whether we stepped on a matching or crossed-over L/R arrow.
          addStep ft L steps = steps ++ [ft == LeftFoot]
          addStep ft R steps = steps ++ [ft == RightFoot]
          addStep ft _ steps = steps -- U/D don't help to determine L/R footing.
          -- remember where this foot ends up (U/D important here for brackets)
          -- XXX: is this wrong? overwriting a fixed (l,r)
          -- with something that could be flipped later
          rememberFoot LeftFoot  arrow (_,r) = ([arrow],r)
          rememberFoot RightFoot arrow (l,r) = (l,[arrow])
          stream s = s { steps = steps s + 1, lastStep = Just arrow,
                         lastFoot = Just foot,
                         stepsLR = addStep foot arrow $ stepsLR s,
                         lastArrowLR = rememberFoot foot arrow $ lastArrowLR s,
                         stepsAndJumps = stepsAndJumps s + 1 }

analyze :: [Step] -> AnalysisState
analyze = commitStream Nothing . foldl analyzeStep (S 0 0 0 0 0 0 0 0 Nothing Nothing False Nothing False [] ([],[]) ([],[]))

-- turns a line of stepchart, eg "0100", into a Step, e.g. "D"
-- in stepchart-ese: 1 = tap; 2 = hold; 3 = hold release; 4 = roll; M = mine
stepify :: String -> Maybe Step
stepify = step . map fst . filter snd . zip [L,D,U,R] . map (flip elem "124")
    where step [] = Nothing
          step [arrow] = Just $ A arrow
          step arrows = Just $ J $ toJump arrows -- $ sort arrows -- sort not nec.
          toJump [L,D] = LD
          toJump [L,U] = LU
          toJump [D,R] = DR
          toJump [U,R] = UR
          toJump [L,R] = LR -- cannot bracket opposite-arrow jumps
          toJump [D,U] = DU -- (perhaps aka, "candle jumps"???)
          toJump [_,_] = error "somehow not sorted??"
          toJump (_:_:_:_) = Other -- any other jumps must be 3+

-- handles chart metadata and formats the analysis result into tsv
process :: (String, [String]) -> String
process (title,(name:diff:feet:_:rest)) = intercalate "\t" $ metadata ++ result
    where -- chop off the trailing ":"
          trim = reverse . tail . dropWhile isSpace . reverse . dropWhile isSpace
          metadata = map trim [title,name,diff,feet]
          finalState = analyze $ catMaybes $ map (stepify . dropWhile isSpace) rest
          -- printing stepsandjumps first for backwards compatibility with the
          -- 2017-era test cases, which expect just 'steps' (== streamy steps)
          -- (their answers must be a uninterrupted substring of the output)
          result = map (show . ($ finalState)) [stepsAndJumps, steps, xovers, switches,
                                                jacks, doubles, xoverfs, brackets]

-- splits up the input file by the "#NOTES:" lines
charts :: String -> [String] -> [(String, [String])]
charts t [] = []
charts t xs = chart ++ charts title rest
    where title = fromMaybe t $ drop 7 <$> dropWhile isSpace <$> find (isInfixOf "#TITLE:") xs
          -- get everything after the first "#NOTES"
          notnotes = not . isInfixOf "#NOTES:"
          stuff = tail $ dropWhile notnotes xs
          -- split this chart from the rest
          c0   = takeWhile notnotes stuff
          rest = dropWhile notnotes stuff
          -- accept only singles charts
          single = isInfixOf "dance-single:"
          chart = if any single c0 then [(title, tail $ dropWhile (not . single) c0)] else []

main = interact $ unlines . map process . charts "" . lines
